File: apps/web/src/app/core/services/notifications.service.ts
import { Injectable } from '@angular/core';
import { environment } from '../../../envirnoments/envirnoment';
import { GqlService } from './gql.service';
import { AuthService } from './auth.service';

export type NotificationActor = {
  user_id: string;
  display_name?: string | null;
  username?: string | null;
  avatar_url?: string | null;
};

export type NotificationItem = {
  id: string;
  user_id: string;
  actor_id?: string | null;
  type: string;
  entity_type?: string | null;
  entity_id?: string | null;
  read_at?: string | null;
  created_at: string;
  actor?: NotificationActor | null;
};

const LIST_NOTIFICATIONS = `
query Notifications($limit: Int, $before: String) {
  notifications(limit: $limit, before: $before) {
    id
    user_id
    actor_id
    type
    entity_type
    entity_id
    read_at
    created_at
    actor {
      user_id
      display_name
      username
      avatar_url
    }
  }
}
`;

const UNREAD_COUNT = `
query NotificationsUnreadCount {
  notificationsUnreadCount
}
`;

const MARK_READ = `
mutation MarkNotificationRead($id: ID!) {
  markNotificationRead(id: $id)
}
`;

const MARK_ALL_READ = `
mutation MarkAllNotificationsRead {
  markAllNotificationsRead
}
`;

@Injectable({ providedIn: 'root' })
export class NotificationsService {
  private endpoint = environment.graphqlEndpoint || 'http://localhost:3000/graphql';

  constructor(
    private gql: GqlService,
    private auth: AuthService
  ) {}

  async list(limit = 40, before?: string | null) {
    try {
      return await this.gql.request<{ notifications: NotificationItem[] }>(LIST_NOTIFICATIONS, {
        limit,
        before: before ?? null,
      });
    } catch {
      return await this.quietRequest<{ notifications: NotificationItem[] }>(LIST_NOTIFICATIONS, {
        limit,
        before: before ?? null,
      }, { notifications: [] });
    }
  }

  async unreadCount() {
    try {
      return await this.gql.request<{ notificationsUnreadCount: number }>(UNREAD_COUNT);
    } catch {
      return await this.quietRequest<{ notificationsUnreadCount: number }>(
        UNREAD_COUNT,
        undefined,
        { notificationsUnreadCount: 0 }
      );
    }
  }

  async markRead(id: string) {
    return this.gql.request<{ markNotificationRead: boolean }>(MARK_READ, { id });
  }

  async markAllRead() {
    return this.gql.request<{ markAllNotificationsRead: number }>(MARK_ALL_READ);
  }

  private async quietRequest<T>(query: string, variables: any, fallback: T): Promise<T> {
    try {
      const token = await this.auth.getAccessToken();
      const res = await fetch(this.endpoint, {
        method: 'POST',
        headers: {
          'content-type': 'application/json',
          accept: 'application/json',
          ...(token ? { authorization: `Bearer ${token}` } : {}),
        },
        body: JSON.stringify({ query, variables: variables ?? {} }),
      });
      const text = await res.text();
      const json = text ? JSON.parse(text) : null;
      if (!res.ok || json?.errors?.length || !json?.data) return fallback;
      return json.data as T;
    } catch {
      return fallback;
    }
  }
}

File: apps/mobile/ios/WorldApp/WorldApp/Services/NotificationsService.swift
import Foundation

@MainActor
final class NotificationsService {
    static let shared = NotificationsService()

    private let gql = GraphQLService.shared
    private init() {}

    func list(limit: Int = 40) async -> [NotificationItem] {
        struct Response: Decodable { let notifications: [GraphQLNotification] }
        let query = """
        query Notifications($limit: Int) {
          notifications(limit: $limit) {
            id user_id actor_id type entity_type entity_id read_at created_at
            actor { user_id display_name username avatar_url }
          }
        }
        """
        do {
            let result: Response = try await gql.authenticatedRequest(query: query, variables: ["limit": limit])
            return result.notifications.map(\.toModel)
        } catch {
            return []
        }
    }

    func unreadCount() async -> Int {
        struct Response: Decodable { let notificationsUnreadCount: Int }
        let query = "query { notificationsUnreadCount }"
        do {
            let result: Response = try await gql.authenticatedRequest(query: query)
            return result.notificationsUnreadCount
        } catch {
            return 0
        }
    }

    func markRead(_ id: String) async throws {
        let mutation = "mutation($id: ID!) { markNotificationRead(id: $id) }"
        let _: EmptyMutation = try await gql.authenticatedRequest(query: mutation, variables: ["id": id])
    }

    func markAllRead() async throws {
        let mutation = "mutation { markAllNotificationsRead }"
        let _: EmptyMutation = try await gql.authenticatedRequest(query: mutation)
    }
}

private struct GraphQLNotification: Decodable {
    let id: String
    let userID: String
    let actorID: String?
    let type: String
    let entityType: String?
    let entityID: String?
    let readAt: String?
    let createdAt: String
    let actor: GraphQLNotificationActor?

    enum CodingKeys: String, CodingKey {
        case id, type, actor
        case userID = "user_id"
        case actorID = "actor_id"
        case entityType = "entity_type"
        case entityID = "entity_id"
        case readAt = "read_at"
        case createdAt = "created_at"
    }

    var toModel: NotificationItem {
        NotificationItem(
            id: id,
            userID: userID,
            actorID: actorID,
            type: type,
            entityType: entityType,
            entityID: entityID,
            readAt: readAt,
            createdAt: createdAt,
            actor: actor?.toModel
        )
    }
}

private struct GraphQLNotificationActor: Decodable {
    let userID: String
    let displayName: String?
    let username: String?
    let avatarURL: String?

    enum CodingKeys: String, CodingKey {
        case userID = "user_id"
        case displayName = "display_name"
        case username
        case avatarURL = "avatar_url"
    }

    var toModel: PostAuthor {
        PostAuthor(
            userID: userID,
            displayName: displayName,
            username: username,
            avatarURL: avatarURL,
            countryName: nil,
            countryCode: nil,
            lastReadAt: nil
        )
    }
}